compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala

```diff
@@ -697,24 +697,79 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
         implicit pos: Position): js.Tree = {
 
-      val trgSym = {
-        if (sym.isClassConstructor) {
-          /* Get the companion module class.
-           * For inner classes the sym.owner.companionModule can be broken,
-           * therefore companionModule is fetched at uncurryPhase.
+      val owner = sym.owner
+      val isNested = owner.isLifted && !isStaticModule(owner.originalOwner)
+
+      val (trgSym, trgTree) = {
+        if (!sym.isClassConstructor && !static) {
+          /* Default getter is on the receiver.
+           *
+           * Since we only use this method for class internal exports
+           * dispatchers, we know the default getter is on `this`. This applies
+           * to both top-level and nested classes.
+           */
+          (owner, js.This()(encodeClassType(owner)))
+        } else if (isNested) {
+          assert(captures.size == 1,
+              s"expected exactly one capture got $captures ($sym at $pos)")
+
+          /* Find the module accessor.
            *
            * #4465: If owner is a nested class, the linked class is *not* a
            * module value, but another class. In this case we need to call the
            * module accessor on the enclosing class to retrieve this.
+           *
+           * #4526: If the companion module is private, linkedClassOfClass
+           * does not work (because the module name is prefixed with the full
+           * path). So we find the module accessor first and take its result
+           * type to be the companion module type.
            */
-          val companionModule = enteringPhase(currentRun.namerPhase) {
-            sym.owner.companionModule
+
+          val outer = owner.originalOwner
+
+          val modAccessor = {
+            val name = enteringPhase(currentRun.typerPhase) {
+              owner.unexpandedName.toTermName
+            }
+
+            outer.info.members.find { sym =>
+              sym.isModule && sym.unexpandedName == name
+            }.getOrElse {
+              throw new AssertionError(
+                s"couldn't find module accessor for ${owner.fullName} at $pos")
+            }
           }
-          companionModule.moduleClass
+
+          val receiver = captures.head
+
+          val trgSym = modAccessor.tpe.resultType.typeSymbol
+
+          val trgTree = if (isJSType(outer)) {
+            genApplyJSClassMethod(receiver, modAccessor, Nil)
+          } else {
+            genApplyMethodMaybeStatically(receiver, modAccessor, Nil)
+          }
+
+          (trgSym, trgTree)
+        } else if (sym.isClassConstructor) {
+          assert(captures.isEmpty, "expected empty captures")
+
+          /* Get the companion module class.
+           * For classes nested inside modules the sym.owner.companionModule
+           * can be broken, therefore companionModule is fetched at
+           * uncurryPhase.
+           */
+          val trgSym = enteringPhase(currentRun.namerPhase) {
```

Comment:
The comment says "at uncurryPhase" but the code says "at namerPhase". Should it be
```suggestion
           * can be broken, therefore companionModule is fetched at
           * uncurryPhase.
           */
          val trgSym = enteringPhase(currentRun.uncurryPhase) {
```
I know that the previous code was already using `namerPhase`, but that seems suspicious to me. Before `exitingPhase(typerPhase)`, relying on anything from the symbol table seems a bit random.

Reply:
I have added a separate commit before the main one.

Reply:
I have added a separate commit before the main one.